atomic_reactor/util.py

```diff
@@ -1672,3 +1672,186 @@ def chain_get(d, path, default=None):
+
+    @property
+    def custom_parent_image(self):
+        return self._custom_parent_image
+
+    @property
+    def source_registry(self):
+        return self._source_registry
+
+    @property
+    def organization(self):
+        return self._organization
+
+    def __len__(self):
+        return len(self._pullable_parents)
+
+    def __setitem__(self, key, item):
+        if not isinstance(key, ImageName):
+            key = ImageName.parse(key)
+
+        if not base_image_is_custom(key.to_str()):
+            if not key.registry:
+                key.registry = self._source_registry
+
+            if self._should_enclose(key):
+                key.enclose(self._organization)
+
+        try:
+            kindex = self._pullable_parents.index(key)
+        except ValueError:
+            raise KeyError(key)
+
+        self._local_parents[kindex] = ImageName.parse(item)
+        logger.info("set parent image '%s' to '%s'", key.to_str(), item)
+
+    def __getitem__(self, key):
+        if not isinstance(key, ImageName):
+            key = ImageName.parse(key)
+
+        if not base_image_is_custom(key.to_str()):
+            if not key.registry:
+                key.registry = self._source_registry
+
+            if self._should_enclose(key):
+                key.enclose(self._organization)
+
+        try:
+            kindex = self._pullable_parents.index(key)
+        except ValueError:
+            raise KeyError(key)
+        return deepcopy(self._local_parents[kindex])
+
+    def __iter__(self):
+        return iter(self.keys())
+
+    def keys(self):
+        return deepcopy(self._pullable_parents)
+
+    def values(self):
+        return deepcopy(self._local_parents)
+
+    def items(self):
+        return zip(self.keys(), self.values())   # pylint: disable=zip-builtin-not-iterating
+
+    def _create_pullable(self):
+        for index, image in enumerate(reversed(self._original_parents)):
+            image_name = ImageName.parse(image)
+            if base_image_is_scratch(image_name.get_repo()):
+                if index == 0:
+                    self._base_from_scratch = True
+                continue
+            if base_image_is_custom(image_name.to_str()):
+                self._custom_parent_image = True
+                if index == 0:
+                    self._custom_base_image = True
+
+            if image_name not in self._pullable_parents:
+                self._pullable_parents.append(image_name)
+                self._local_parents.append(None)
+
+    def set_source_registry_and_org(self, source_registry, organization=None):
+        """
+        set source registry and organization
+        has to be done before using with expectation of having
+        already source registry and organization,
+        set in  reactor_config, and in flatpak_create_dockerfile
+        """
+        self._source_registry = source_registry
+        self._organization = organization
+        self._enclose_pullable()
```

Comment:
`set_source_registry_and_org()` is expected to be used at most once, right? If I'm not mistaken, using it a second time would kinda break everything.

Could `source_registry` and `organization` be passed in `__init__()` instead? Or are they not available yet at some point where this class is used?

Reply:
it is expected to be called once as there isn't any other reason to call it with different source & organization
it wouldn't exactly break things, but there isn't reason for it

class is created in inner.py with set_df_path and for flatpak set_df_path is called in flatpak_dockerfile_plugin

source & organization is known in reactor_config plugin